src/Storage/Sha1FileHashingService.php

```diff
@@ -0,0 +1,193 @@
+        $fullHash = hash_final($hc);
+
+        return $fullHash;
+    }
+
+    /**
+     * Valid hashing algorithm constant that can be passed to `hash_init`.
+     * @note A sub class could override this to create to use an alternative algorithm.
+     * @return string
+     */
+    protected function algo()
+    {
+        return 'sha1';
+    }
+
+    /**
+     * Get the matching Filesystem object
+     * @param string|Filesystem $fs
+     * @throws InvalidArgumentException
+     * @return Filesystem
+     */
+    private function normaliseFilesystem($fs)
+    {
+        if ($fs instanceof Filesystem) {
+            return $fs;
+        }
+
+        if (!in_array($fs, [AssetStore::VISIBILITY_PUBLIC, AssetStore::VISIBILITY_PROTECTED])) {
+            throw new InvalidArgumentException(
+                'Sha1FileHashingService: $fs must be an instance of Filesystem or an AssetStore VISIBILITY constant.'
+            );
+        }
+
+        $filesystem = Injector::inst()->get(sprintf('%s.%s', Filesystem::class, $fs));
+        return $filesystem;
+    }
+
+    public function compute($fileID, $fs)
+    {
+        if ($hash = $this->get($fileID, $fs)) {
+            return $hash;
+        }
+
+        $fs = $this->normaliseFilesystem($fs);
+        $stream = $fs->readStream($fileID);
+        $hash = $this->computeStream($stream);
+
+        $this->set($fileID, $fs, $hash);
+
+        return $hash;
+    }
+
+    public function compare($hashOne, $hashTwo)
+    {
+        // Empty hash will always return false, because they are no validatable
+        if (empty($hashOne) || empty($hashTwo)) {
+            throw new InvalidArgumentException('Sha1FileHashingService::validateHash can not validate empty hashes');
+        }
+        // Return true if $hashOne start with $hashTwo or if $hashTwo starts with $hashOne
+        return strpos($hashOne, $hashTwo) === 0 || strpos($hashTwo, $hashOne) === 0;
+    }
+
+    public function isCached()
+    {
+        if ($this->cachable === null) {
+            $this->cachable = self::config()->get('default_cachable');
+        }
+
+        return $this->cachable;
+    }
+
+    public function enableCache()
+    {
+        $this->cachable = true;
+    }
+
+    public function disableCache()
+    {
+        $this->cachable = false;
+        $this->flush();
+    }
+
+    /**
+     * @return CacheInterface
+     */
+    private function getCache()
+    {
+        if (!$this->cache) {
+            $this->cache = Injector::inst()->get(
+                sprintf('%s.%s', CacheInterface::class, 'Sha1FileHashingService')
+            );
+        }
+        return $this->cache;
+    }
+
+    /**
+     * Build a unique cache key for the provided parameters
+     * @param string $fileID
+     * @param string|Filesystem $fs
+     * @return string
+     */
+    private function buildCacheKey($fileID, $fs)
+    {
+        $fs = $this->normaliseFilesystem($fs);
+        $fsID = spl_object_hash($fs);
```

Comment:
I've confirmed that those hashes are the same across PHP executions, otherwise a filesystem based cache would be kinda pointless - see https://www.php.net/manual/en/function.spl-object-hash.php#76220

Reply:
Nothing to do here.